sopel/tools/memories.py

```diff
@@ -149,14 +165,37 @@ def __init__(
         *args,
         identifier_factory: IdentifierFactory = Identifier,
     ) -> None:
-        super().__init__(*args)
+        if len(args) > 1:
+            raise TypeError(
+                'SopelIdentifierMemory expected at most 1 argument, got {}'
+                .format(len(args))
+            )
+
         self.make_identifier = identifier_factory
         """A factory to transform keys into identifiers."""
 
+        if len(args) == 1:
+            super().__init__(self._process_arg(args[0]))
+        else:
+            super().__init__()
+
     def _make_key(self, key: Optional[str]) -> Optional[Identifier]:
-        if key is not None:
-            return self.make_identifier(key)
-        return None
+        if key is None:
+            return None
+        return self.make_identifier(key)
+
+    def _process_arg(self, maybe_mapping=None):
```

Comment:
> I now can't think of any reason not to just always return `Generator[tuple[Identifier, Any], None, None]`. Thoughts?

I'm pretty luke-warm on `Generator[T, None, None]` and usually opt for `Iterable[T]` instead since it conveys about the same information, but I agree that it makes sense for this internal helper to return a consistent type, and doing a lazy stream of converted-key/val pairs makes sense.